x-pack/qa/full-cluster-restart/build.gradle

```diff
@@ -108,7 +109,12 @@ for (Version bwcVersion : bwcVersions.indexCompatible) {
     useCluster testClusters."${baseName}"
     dependsOn "${baseName}#oldClusterTest"
     doFirst {
-      testClusters."${baseName}".goToNextVersion()
+      testClusters."${baseName}".goToNextVersion({ ->
+        if (Version.fromString(bwcVersionString).onOrAfter("8.0.0")) {
+          // verify that on-disk metadata can be read using command-line tools
+          testClusters."${baseName}".runElasticsearchBinScriptWithInput("y", "elasticsearch-node", "read-and-write-metadata")
```

Comment:
Is this meant to be a test of the elasticsearch-node tool? If so it seems like  it better belongs as an actual test project for elasticsearch-node, rather than conflating it into cluster restart tests.

Reply:
In contrast to the unit tests that we also have in this PR (and the existing more "integration-type" tests of the tools), this is more of an end-to-end test that makes sure that the `preserve_customs` metadata format indeed works for any real cluster state that we have. The reason why it is inlined in this test is that this test has a cluster with all kinds of custom metadata in it (it starts many x-pack components), which makes it very effective at detecting any kind of problems with the serialization format. If we had a dedicated test for this (which is what the unit tests already do), we could potentially miss bugs. While I generally prefer not to mix different concerns in one place, the opportunity provided by this test to have a large selection of the real metadata is too good to pass. An extra test project would not provide any benefit over the existing tests that we already have. For that extra project to be as effective as this test here, it would have to start / run any and all x-pack components, putting a lot of our actual metadata into the cluster state. This would be a lot of boilerplate. In that case I would prefer not to have that kind of test at all. Given that this is a tiny piece in this full-cluster-restart test, I would prefer to keep this, but can also drop it if that's what you prefer.

Reply:
I have two concerns with it here. The first is practical: if this fails, most people on the team won't understand what is failing, and will think it is related to bwc being broken (which may be the case, but is *very* unlikely related to any actual bwc change they are making, since they would not be touching this code). The second reason is more theoretical, in that often when bwc tests fail, the failure is attributed to gradle's fault simply because a tool failed (or Elasticsearch failed to startup). This adds a non inconsequential cognitive load to the build team's responsibilities for test triage. Keeping tests isolated helps better route failures to the right area team for analysis.

Based on your description, it sounds like you are using this as a sort of randomized test. While I see value in ensuring the tool works at a system level (for example in the packaging tests) in a basic way, I'm not sure what overlapping in unrelated tests would catch here. If I understand the code correctly, it is meant to be completely agnostic to the particular custom metadata, so what deficiency is there in doing this in a more isolated way? What could we miss?

Reply:
I've removed the end-to-end x-pack test. In contrast to the earlier approach (loading plugins in the tool), the tests should now cover customs well enough.